models/badges.js

```diff
@@ -1,62 +1,132 @@
+const admin = require("firebase-admin");
 const firestore = require("../utils/firestore");
 const badgeModel = firestore.collection("badges");
-const { fetchUser } = require("../models/users");
+const userBadgeModel = firestore.collection("userBadges");
+const {
+  convertFirebaseTimestampToDateTime,
+  convertFirebaseDocumentToBadgeDocument,
+  assignUnassignBadgesInBulk,
+} = require("../utils/badges");
+const { chunks } = require("../utils/array");
+const { DOCUMENT_WRITE_SIZE, ERROR_MESSAGES } = require("../constants/badges");
+const MODEL_ERROR_MESSAGES = ERROR_MESSAGES.MODELS;
 
 /**
  * Fetches the data about our badges
  * @param query { Object }: Filter for badges data
- * @return {Promise<badgeModel|Array>}
+ * @return {Promise}: <badgeModel|Array> returns all badges
  */
 const fetchBadges = async ({ size = 100, page = 0 }) => {
   try {
     const snapshot = await badgeModel
       .limit(parseInt(size))
       .offset(parseInt(size) * parseInt(page))
       .get();
-
-    const allBadges = [];
-    snapshot.forEach((doc) => {
-      allBadges.push(doc.data());
-    });
-    return allBadges;
+    // INFO: timestamp to date time logic surfaced from
+    // https://stackoverflow.com/a/66292255
+    return snapshot.docs.map((doc) => convertFirebaseDocumentToBadgeDocument(doc.id, doc.data()));
   } catch (err) {
-    logger.error("Error retrieving badges", err);
-    return err;
+    logger.error(MODEL_ERROR_MESSAGES.FETCH_BADGES, err);
+    throw Error(err?.message ?? MODEL_ERROR_MESSAGES.FETCH_BADGES);
   }
 };
 
 /**
  * Fetches the data about user badges
- * @param query { Object }: Filter for badges data
- * @return {Promise<userBadgeModel|Array>}
+ * @param userId <string>: Filter for badges data
+ * @return {Promise}: <{badges: Array<badge>} returns badges array
+ */
+async function fetchUserBadges(userId) {
+  try {
+    const badgeIdsSnapshot = await userBadgeModel.where("userId", "==", userId).get();
+    // INFO: if userId is incorrect it returns success response
+    if (badgeIdsSnapshot.empty) {
+      return { badges: [] };
+    }
+    const badgeDocReferences = badgeIdsSnapshot.docs.map((doc) => {
+      const badgeId = doc.get("badgeId");
+      return firestore.doc(`badges/${badgeId}`);
+    });
+    // INFO: getAll accepts unpacked array
+    // TODO: check getAll limitiations
+    const badgesSnapshot = await firestore.getAll(...badgeDocReferences);
+    const badges = badgesSnapshot.map((doc) => convertFirebaseDocumentToBadgeDocument(doc.id, doc.data()));
+    return { badges };
```

Comment:
why return an object, can't we just return the output array as it is, or is there anything else?

Reply:
It is left by mistake, previously I was sending more info but later 
decided not to send as that will be unneccessary info. 
I will remove in next iteration or when adding rate-limiting middelware.